### PR TITLE
Add Embroider support

### LINKED ISSUE #30: Is Embroider support on the roadmap?
We have an application that uses this helper. After getting past some other issues, we were able to build the app with Embroider. Then we turned on `staticHelpers`, and builds were failing until we removed references to `(element ...)`. One was a reference to `(element this.tag)` and the other was `(element '')`.

Is this known? Is Embroider support on the roadmap?

### PR DESCRIPTION
**Pull request by simonihmig**
Fixes #30 

Embroider has a [special dependency rule](https://github.com/embroider-build/embroider/blob/4fad67f16f811e7f93199a1ee92dba8254c42978/packages/compat/src/addon-dependency-rules/ember-element-helper.ts) in place in to support the dynamic component pattern in Embroider, which made this addon work in Embroider *for a while*. However it seems https://github.com/embroider-build/embroider/pull/608 added some stricter checks for dynamic component usage that again made this addon fail under Embroider.

This should now hopefully solve the issue once and for all, by wrapping the private `-element` helper in Embroider's [`ensure-safe-component`](https://github.com/embroider-build/embroider/tree/master/packages/util#ensuresafecomponent), which also allows us to get rid of the `-dynamic-element-*` components (see code comment).

Tested this locally in an app, and added ember-try scenarios.

To my understanding the changes here should be ok, especially the use of the `ensure-safe-component` helper, but @ef4 if you have the time to give this a review from Embroider's perspective, this would be helpful!

### CODE REVIEW

File: tests/integration/helpers/element-test.js
Comment (simonihmig) [inline tests/integration/helpers/element-test.js:92]: I removed this as Embroider seems to have its own `hbs` babel-plugin, that [throws](https://github.com/embroider-build/embroider/blob/master/packages/core/src/babel-plugin-inline-hbs.ts#L123) when receiving more than one argument. I am not sure what this option does here, so please advice if/how this needs to be fixed!?

**Comment by simonihmig** [comment]
A few scenarios are failing, but it *seems* for unrelated reasons. Also the latest scheduled run on `main` had some [failures](https://github.com/tildeio/ember-element-helper/actions/runs/367087755)...

**Comment by ef4** [comment]
Hmm, if you're able to delete the `-dynamic-element` components that seems like `ensure-safe-component` maybe isn't stable enough, and it could be hurting re-render performance.

**Comment by simonihmig** [comment]
> Hmm, if you're able to delete the -dynamic-element components that seems like ensure-safe-component maybe isn't stable enough, and it could be hurting re-render performance.

I don't follow you tbh... just to prevent any misunderstandings: the -dynamic-element components are not used anymore in this PR. They are just still present as Embroider's dependency rule that I added a while ago still [refers to them](https://github.com/embroider-build/embroider/blob/master/packages/compat/src/addon-dependency-rules/ember-element-helper.ts#L7), and removing them here at this point in time would make Embroider complain about those missing.

**Comment by simonihmig** [comment]
> A few scenarios are failing, but it seems for unrelated reasons.

Hm, not quite (re: unrelated):
* yes, classic and jquery also fail on the main branch
* beta and canary fail with [`owner.factoryFor` being undefined](https://github.com/tildeio/ember-element-helper/pull/34/checks?check_run_id=1413058303#step:10:20), the call stack includes `ensureSafeComponent`. Which has some [special workaround for 3.24](https://github.com/embroider-build/embroider/blob/master/packages/util/addon/ember-private-api.js#L30-L31) (which seems to have been promoted to beta recently), so maybe that is not working correctly? /cc @ef4 
* 3.4 fails with [Element modifiers are not allowed in components](https://github.com/tildeio/ember-element-helper/pull/34/checks?check_run_id=1413059049) which I am not sure if/how this is related to my changes 🤔

File: addon/helpers/-element.js
Comment (ef4) [inline addon/helpers/-element.js:23]: This is going to give you a different component instance every time this helper recomputes. It's not stable. *Everything* below this point in the render hierarchy will get destroyed and re-created.

Instead, we really only want it to change when the `tagName` argument changes.

Comment (ef4) [reply addon/helpers/-element.js:23]: Oh, I see that higher in the code this is guarded by `tagName !== this.tagName`, so I think my above comment is wrong.

Comment (simonihmig) [reply addon/helpers/-element.js:23]: > so I think my above comment is wrong.

I hope so! :)

However your comment made me aware and worry a bit that we have the `.extend()` in the hot path here. Reusing the same component class every time does not work though when the tag name changes, as the component instance is not invalidated, I guess for the same reason there were two separate -dynamic-element components before. But maybe we should also return two alternating component classes, also because the `ensure-safe-component` helpers will register a new component each and every time IIUC: https://github.com/embroider-build/embroider/blob/master/packages/util/addon/index.js#L45-L53

Not sure why anyone would change tag names for the same instance, but I guess we can still cover this case better?

**Comment by ef4** [comment]
@simonihmig I [confirmed](https://github.com/embroider-build/embroider/pull/623) that there's no stability problem in ensure-safe-component, so my fears above were unfounded.

Also, the problem you hit with ember beta is fixed by https://github.com/embroider-build/embroider/pull/620 and will be released soon.

File: addon/helpers/-element.js
Comment (ef4) [reply addon/helpers/-element.js:23]: Yes, I think it would be good to keep two alternating classes. The ensureSafeComponent function definitely makes that easier now, since they can easily be defined inline in this file, instead of being globally exposed components.

Comment (simonihmig) [reply addon/helpers/-element.js]: In the second commit I added back the alternating component class pattern, however as you said defined locally without exposing them.

A bunch of tests fail though, and the interesting part here is that *all* of them pass when running in isolation. I think the reason is that [this WeakMap)(https://github.com/embroider-build/embroider/blob/master/packages/util/addon/index.js#L42) survives across tests, although the owner changes. So any following test would reuse the map, assume that the component from the previous test run is already registered with the owner, but which is not the case as the owner changed.

I prepared a PR for embroider, that fixes this problem locally (when using it with `yarn link`): https://github.com/embroider-build/embroider/pull/624

**Comment by simonihmig** [comment]
I updated the Embroider dependencies after its latest release, which makes most scenarios pass.
The still failing ones are:
* default-jquery 
* classic
* 3.4

The former two are unrelated to this PR, and [failing with the same error on `main` due to `set-env` not working anymore](https://github.com/tildeio/ember-element-helper/runs/1445122033?check_suite_focus=true#step:6:40)

The last one is tricky. Seems the second argument (`{ insertRuntimeErrors: true }`) to `hbs` [mentioned above](https://github.com/tildeio/ember-element-helper/pull/34#discussion_r525256246) is there to allow Ember <3.11 to compile the template. While at the same time this second argument makes Embroider throw a compilation error as mentioned.

I already tried replacing `ember-compatibility-helpers` with `@embroider/macros`, in the hope that the latter's dead branch elimination would make the offending `hbs` call be hidden from the template compiler. But it seems the template compiler's transform happens before that of `@embroider/macros`, so that doesn't help here. 

So as it stand either the Ember 3.4 scenario or the two Embroider scenarios added here will fail with a build error. I have no better idea to solve this than to disable the 3.4 scenario!? 🤔

However we end up fixing this, it's certainly "just" a CI issue. Except for this, I believe the changes here should be fine now from my PoV, and looking forward for a final review! (as this is a blocking issue for some addon's embroider progess :wink: )
/cc @chancancode

**Comment by ef4** [comment]
@simonihmig I think the solution here is that Embroider needs to implement `insertRuntimeErrors`.

**Comment by simonihmig** [comment]
> I think the solution here is that Embroider needs to implement insertRuntimeErrors.

I planned to do that, but time was flying away, and as it stands I am not able to do that in the near future, at least not in the remaining days of this year.

As stated above, the issue is "just" about testability, not the actual code. So is there a chance we can land this without being blocked on the test coverage here (again, the change in the actual code is blocking embroider adoption for some addons, at least `ember-bootstrap`)? @chancancode

**Comment by ef4** [comment]
insertRuntimeErrors is now implemented here https://github.com/embroider-build/embroider/pull/658

**Comment by ef4** [comment]
Embroider 0.35.0 was released with support for insertRuntimeErrors.

**Comment by simonihmig** [comment]
> Embroider 0.35.0 was released with support for insertRuntimeErrors.

Awesome, thanks so much! 🙏

I previously replaced ember-compatibility-helpers with @embroider/macros, hoping to fix the insertRuntimeErrors issue by that, which did not work out. But I figured this change is useful by itself, so I pushed that here. Also Ember 3.25 Canary started to throw a different error for that "invalid usages: it does not take a block"  test (related to [this GlimmerVM update](https://github.com/emberjs/ember.js/pull/19278)), so I also pushed a commit fixing that.

CI is "green" now (again, 2 failing jobs are broken on main branch also), so ready for a final review please! 🙏

**Comment by simonihmig** [comment]
As the CI failures on the main branch have been fixed meanwhile, I rebased this, and CI is green now!

Hope we can finally get this reviewed and eventually merged, to unblock Embroider adoption in the ecosystem! @chancancode @rwjblue

File: lib/element-helper-syntax-plugin.js
Comment (rwjblue) [inline lib/element-helper-syntax-plugin.js:54]: Hmm. Where does this come from when the app is not being built by embroider? (`@embroider/macros` is only a devDep)

File: package.json
Comment (rwjblue) [inline package.json:50]: This seems somewhat odd, why drop this down to 3.4?

File: lib/element-helper-syntax-plugin.js
Comment (simonihmig) [reply lib/element-helper-syntax-plugin.js:54]: This is provided by `@embroider/util`, which is a "real" dependency!

File: package.json
Comment (simonihmig) [reply package.json]: Oh, good catch! Don't remember exactly, been a while. But I guess I just downgraded locally to investigate something failing, and accidentally commited that. Just reverted that change!

**Comment by chriseppstein** [comment]
Is there an ETA on landing this?

**Comment by simonihmig** [comment]
> Is there an ETA on landing this?

Only waiting for final approval. @rwjblue I addressed your feedback long ago, could you please have another look?

**Comment by cibernox** [comment]
what's the status of this?

**Comment by simonihmig** [comment]
> what's the status of this?

Same as 3 weeks ago:

> Only waiting for final approval.

🤷‍♂️

**Comment by chancancode** [comment]
We also needed to fix our embroider build at Skylight so we can look into merging and tagging it this week.

I am not sure I am the best person to catch mistakes relating to embroider itself, other than confirming that it does indeed work in the app. Anything you would like to call out specifically?

Sorry for the wait!

**Comment by simonihmig** [comment]
> I am not sure I am the best person to catch mistakes relating to embroider itself, other than confirming that it does indeed work in the app. Anything you would like to call out specifically?

Not really. @ef4 and @rwjblue both did an initial review, and I think I have addressed everything. Also I am using this branch in an Embroider-built app, and this PR adds test coverage for Embroider, so I am pretty confident it does work as it should.

File: package.json
Comment (rwjblue) [inline package.json:24]: Sorry, mind bumping this to latest (0.39.1)?

File: package.json
Comment (rwjblue) [inline package.json:31]: also should go to 0.39.1

**Review by rwjblue — COMMENTED** [review]
OK, sorry for the long delay again @simonihmig. I think the only thing remaining here is to bump the @embroider packages to latest, then we are good to go.

**Comment by simonihmig** [comment]
> I think the only thing remaining here is to bump the @embroider packages to latest, then we are good to go.

@rwjblue rebased and updated all embroider dependencies!

**Comment by rwjblue** [comment]
Thanks, sorry again for the delays.

**Review by rwjblue — APPROVED** [review]

**Comment by chancancode** [comment]
I can tag/publish – is this a minor or major bump?

**Comment by chancancode** [comment]
I guess we are pre-1.0 so I will push a 0.5.0

**Comment by rwjblue** [comment]
@chancancode - It should be fully safe (aka non-breaking)

**Comment by chancancode** [comment]
Oops, pushed 0.5.0 already. I supposed it could have been 0.4.1 but doesn't really matter that much

**Pull request merged**
